Add cast app media

eerovil opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add a new handler for the `media_player.play_media` service for `media_content_type: cast`. It allows one to cast supported media using native chromecast applications.

More support for apps can be added either via [Cast Extensions](https://github.com/Eerovil/castextensions) or to pychromecast. Some apps require heavier dependencies and manual configuration, so those can't be integrated into pychromecast. 


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: home-assistant/home-assistant.io#13269

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

eerovil committed:
Add new service: cast_app

This service allows you to cast any pychromecast supported app
and start playing media.

eerovil committed:
Temporarily use my fork of pychromecast

eerovil committed:
Don't set a default in optional parameter

eerovil committed:
Try to import additional cast extensions

eerovil committed:
Use cast extensions as primary lib

eerovil committed:
Use new location for quick_play

eerovil committed:
Revert "Temporarily use my fork of pychromecast"

This reverts commit 27087e2d2b08dc02e3801d2aa6e26198565ba458.

eerovil committed:
Add service info

homeassistant commented:
Hi @Eerovil, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#35114).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @Eerovil, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#35114).

Once you do that we will be able to review and accept this pull request.

Thanks!

Eerovil commented:
The remote build will not pass since pychromecast has not yet been updated with my proposed changes. This PR cannot be merged before pychromecast is update, but it could be reviewed.

frenck commented:
@Eerovil Can you please convert the PR to a draft until it is possible to merge it? thanks 👍

emontnemery commented:
This is a good idea, but instead of adding a new service, do special handling for `media_type=cast`, in the same manner as for plex:
https://github.com/home-assistant/core/blob/7f2c6b43d2e555eeb90ae226344c43a8d4209fd4/homeassistant/components/plex/media_player.py#L566-L571

In the cast of cast, media_id should be valid JSON with at least `app_id` and `content_id` present.

emontnemery commented on homeassistant/components/cast/media_player.py:
```diff
@@ -479,6 +502,30 @@ def play_media(self, media_type, media_id, **kwargs):
+        try:
+            from importlib import import_module
+
+            extended_quick_play = import_module("castextensions").quick_play
```
Make `castextenstions` a dependency of either `pychromecast` or `homeassistant.components.cast` instead.

Eerovil replied:
Cast extensions has some large dependencies, such as AndroidViewClient. Isn't it a bad idead to add such a dependency to all cast integration users?

emontnemery replied:
OK, that's no good.
Can you remove the cast extensions part from this PR? It would be better suited as a custom component.

Eerovil replied:
Sure, I'll remove it.

But I'm not exactly sure how to get similar functionality with a custom component. Should I try to extend the cast-component or something like that?

emontnemery replied:
A really simple way would be to fork `cast/media_player.py` and use that as a custom component. It will then override the core component.
Not exactly elegant though.

Not sure how to extend the cast-component otherwise, Ideally you'd want to import from `cast/media_player.py` and only override `play_media`.

Eerovil replied:
It appears I would need to completely fork `cast` and replace it with `cast_extended` (or something). Doesn't seem ideal.

Eerovil commented:
Extending play_media sounds good. I'll do that!

eerovil committed:
Don't create a new service, extend media_play instead

eerovil committed:
Don't try to import cast extensions

MartinHjelmare commented on homeassistant/components/cast/media_player.py:
```diff
@@ -476,6 +478,19 @@ def media_seek(self, position):
+                    return
+                except NotImplementedError:
+                    _LOGGER.error("App %s not supported", app_name)
+
```
We want to return here after the error log, I think.

Eerovil replied:
Hmm, I figured we shoild try regular media casting on failure.

But maybe you're right, it should just fail since the user did choose the `media_content_type: cast` option.

eerovil committed:
Don't continue after errors.

Eerovil commented:
I managed to get good results with Cast Extensions as a custom component now. The component also implements `cast_extensions.media_play` in the same way, but the problem is interfacing with the correct Chromecast. I can fetch the chromecast name from the entity registry, but doing it like that causes a big delay (I need to fetch all chromecasts, check the names to get the chromecast instance).

I'm thinking I could create a signal listener here for `CastDevice` which would allow me to send arbitrary events to the chromecast. Does that seem like a good idea? Any ideas are welcome.

MartinHjelmare commented:
Ok. @emontnemery?

Eerovil commented:
Something like this: ```

    def _handle_signal_cast_function(
        self,
        entity_id: str,
        action: str,
        data: dict,
    ):
        """Handle a cast signal. To stop, just use service turn_off"""
        if entity_id != self.entity_id:
            return

        if action == "quick_play":
            app_name = data.pop("app_name")
            try:
                quick_play(self._chromecast, app_name, data)
            except NotImplementedError:
                _LOGGER.error("App %s not supported", app_name)
        elif action == "start_app":
            self._chromecast.start_app(data['app_id'])
```
start_app is needed for AndroidViewClient: To get instant UI feedback we can start the app instantly, then let the android device do the rest later.

eerovil committed:
Add signal listener for cast extensions

eerovil committed:
Fix both handlers using same connection

emontnemery commented on homeassistant/components/cast/media_player.py:
```diff
@@ -476,8 +486,20 @@ def media_seek(self, position):
+            except NotImplementedError:
+                _LOGGER.error("App %s not supported", app_name)
+        else:
+            # We do not want this to be forwarded to a group
```
Move this comment before `if media_type == CAST_DOMAIN:`, it's a reminder that `play_media` will be interpreted as always targetting the specifi cast, unlike media controls which may be forwarded to a speaker group

eerovil committed:
Move comment.

It's a reminder that play_media will be interpreted as always targetting the specific cast, unlike media controls which may be forwarded to a speaker group.

emontnemery commented:
@Eerovil I think this is good to go after you bump pychromecast to 5.1.0

eerovil committed:
Bump pychromecast to support quick_play

balloob commented on homeassistant/components/cast/media_player.py:
```diff
@@ -218,6 +222,9 @@ async def async_added_to_hass(self):
         self._cast_view_remove_handler = async_dispatcher_connect(
             self.hass, SIGNAL_HASS_CAST_SHOW_VIEW, self._handle_signal_show_view
         )
+        self._cast_function_remove_handler = async_dispatcher_connect(
```
I don't see anything using this?

Eerovil replied:
`_will_remove_from_hass`

Or is that correct usage?

emontnemery replied:
@eerovil I think the question is who will send `SIGNAL_HASS_CAST_APPLICATION`

Eerovil replied:
Oh, right. It's for extending the functionality. Cast extensions sometimes requires just launching a cast app by its ID. 

The signal isn't needed if the service would also support launching an arbitrary application (i.e. not supported by pychromecast, say netflix). Maybe I could add a possibility to input an "app_id" in place of "app_name"?

emontnemery replied:
Yes, that sounds like a simpler solution than adding the signal stuff.

However, is it useful to just start an app?
Don't you need a way to send messages to the app also?

Eerovil replied:
For apps using Android remote control for casting, the app is first via pychromecast for better UX. Otherwise there is no feedback for the action until ~15 seconds.

That's the only use case currently.

balloob replied:
Integrations should only offer external control via services or via functions in `__init__.py` from the integration. We can't keep track of random signal that might or might not be used by other integrations.

Eerovil replied:
So, this is now resolved, as I moved the functionality into the service?

eerovil committed:
Support launching an arbitrary cast application via service

eerovil committed:
Delete left over dispatcher remove handler

emontnemery commented:
I think this is good to go, can you please open a documentation PR also?

Maybe some inspiration here: https://www.home-assistant.io/integrations/plex#media-player
No need to document the different supported apps, but mention the structure of `media_id` and how it will be used.

Eerovil commented:
I made one here: https://github.com/home-assistant/home-assistant.io/pull/13269

Perhaps it's a bit too in-depth.
